core/src/solution_submission/retry.rs

```diff
@@ -99,9 +106,44 @@ fn gas_price(estimated_price: U256, price_increase_count: u32, cap: U256) -> U25
     cap.min(U256::from(new_price as u128))
 }
 
+/// The output type of futures used in `select_all`.
+enum SelectFutureOutput {
+    SolutionSubmission(Result<(), MethodError>),
+    Timeout,
+}
+
+// Return the first future result, skipping nonce errors unless the nonce error is the last result.
+//
+// `futures` contains all solution submission attempts and optionally a timeout. Unfinished futures
+//  stay in the vector when the function returns.
+async fn first_result_ignoring_nonce_errors(
+    futures: &mut Vec<impl Future<Output = SelectFutureOutput> + Unpin>,
+) -> SelectFutureOutput {
+    loop {
+        match future::select_all(futures.split_off(0)).await {
```

Comment:
To avoid panicking in case the vector is empty we can maybe lift the `remaining_future.is_empty()` condition above here (might also make the logic inside the match statement a bit more linear).

I see that we don't have access to the result in that case, but we could create a NonceError (or another error type) in that case since this should really not happen, or should it?

Reply:
Don't like this. We would be handling a case that cannot happen. But I'll add a comment to the function saying the vector can't be empty.

Reply:
In general I prefer panics over errors for logic errors that should never happen and are independent of any outside influence (such as a misbehaving node) as in this case.

Something that I like as well is a `debug_assert!(!futures.is_empty())` as a self-documenting assertion on an invariant that should hold in the function.